DOC: Clarify tiny/xmin in finfo and machar (gh-16253)

rk-mlu opened the pull request:
This commit adds a note to the documentation of finfo clarifying that
the attribute `tiny` does not represent the smallest positive usable
number. Instead, it refers to the smallest positive number with there
being no leading 0's in the mantissa, that is, the smallest positive 
_normalized_ floating-point number.
Alternatively, one can say that `tiny` is the smallest positive floating 
point number with full precision. The smallest positive _subnormal_
floating point number is typically orders of magnitudes smaller, but 
has reduced precision since leading 0's in the mantissa are used to
allow smaller exponents than indicated by `minexp`.

The commit also adds a brief clarification to the docstring of machar.

Closes #16252

rk-mlu committed:
DOC: Modified documentation of finfo and machar

This commit adds a note to the documentation of finfo clarifying that
the attribute `tiny` does not represent the smallest positive usable
number. Instead it refers to the smallest positive number with there
being no leading 0's in the mantisse (normalized floating point number).
Alternatively, `tiny` represents the smallest positive floating point
number with full precision. The smallest positive subnormal floating
point number is typically orders of magnitudes smaller, but has reduced
precision since leading 0's in the mantisse are used to allow smaller
exponents than indicated by `minexp`.

The commit also adds a brief clarification to the docstring of machar.

Closes #16252

rossbar commented on numpy/core/getlimits.py:
```diff
@@ -354,6 +355,15 @@ class finfo:
 
     Notes
     -----
+    As in the IEEE-745 standard, NumPy floating point types use subnormal
+    numbers to fill the gap between 0 and `tiny`. This is achieved by 
+    using bits of the mantissa to allow the representation of smaller 
+    exponents than `minexp`. For example, for 64-bit binary floats in the
+    IEEE-754 standard, the smallest positive value is actually 2**-1074.
+    However, since the number of bits available for the mantissa is 
+    reduced the precision of subnormal floating numbers is also often 
+    reduced significantly.
+    
```
I'd recommend moving this to after the next paragraph "For developers of NumPy..." or, better yet, replacing or significantly reducing the text and providing an external reference.

rk-mlu replied:
Yes, I see your point. I have no objections in moving the paragraph below the next paragraph, of course. I will also think about a shorter version that points to the wiki page.

rossbar replied:
Just an added bit of info you may find helpful: the [numpydoc standard](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard) describes how to add references to external docs etc.

rossbar reviewed (commented):
This is a nice clarification for the docs; however, it might be worth considering replacing some of the text that has been added with an external reference. The descriptions are valuable and interesting, but a lot of the details are not NumPy-specific and thus a more brief summary followed by a reference might be more appropriate. IMO the wiki page on [denormal numbers](https://en.wikipedia.org/wiki/Denormal_number) provides a nice overview of some of the details that have been added here.

rk-mlu committed:
Update machar.py

rk-mlu committed:
Update getlimits.py

More concise version of the clarification in finfo, as recommended by @rossbar.

anirudh2290 commented on numpy/core/getlimits.py:
```diff
@@ -359,6 +360,13 @@ class finfo:
     repeatedly inside your functions is not a problem.
 
+    Note that `tiny` is not actually the smallest positive representable
+    number in a NumPy floating point type. As in the IEEE-745 standard,
```
```suggestion
    number in a NumPy floating point type. As in the IEEE-754 standard,
```

rk-mlu replied:
@anirudh2290  Many thanks!

rk-mlu committed:
Update numpy/core/getlimits.py

Co-authored-by: Anirudh Subramanian <anirudh2290@apache.org>

rossbar commented on numpy/core/getlimits.py:
```diff
@@ -337,8 +337,9 @@ class finfo:
     tiny : float
-        The smallest positive usable number.  Type of `tiny` is an
-        appropriate floating point type.
+        The smallest positive floating point number with there being no
+        leading 0's in the mantissa. Type of `tiny` is an appropriate 
+        floating point type.
```
```suggestion
        The smallest positive non-subnormal (see Notes) floating point
        number. Type of `tiny` is an appropriate floating point type.
```
This is just my opinion, but I have a slight preference for leaving the details of denormal numbers (i.e. "being no leading 0's in the mantissa") out of the parameter description and instead refer to greater levels of detail outside of the parameters section.

LMK what you think!

rk-mlu replied:
@rossbar That is fine with me, too.

Please take note that I used the phrase "being no leading 0's in the mantissa" since it already appeared in the description of another parameter in the same docstring. Cf. `minexp`.

To be keep it consistent, I guess, we should modify the phrase there as well. What do you think?

seberg replied:
Yeah, I was suprised how often we used "mantissa" since I am not sure its a good start for a general audience. I liked the "with full precision", but overall, I guess a see notes is just the right way to explain it.

Mainly commenting, because it seemed to me like we can just drop the current second sentence, saying that it gives out the correct type seems pretty irrelevant (all of these are obviously returned as the same type as is being queried).

mattip replied:
As for "mantissa" - I think we should prefer ["significand"](https://en.wikipedia.org/wiki/Significand). See also this [stronger opinion](https://en.wikipedia.org/wiki/Talk:Significand) against mantissa

rk-mlu replied:
Okay, I tried to combine these comments in the most recent commit. In particular, I tried
to avoid usage of the word mantissa if possible.

Regarding the matter "significand" vs. "mantissa": I have no strong feelings about this. The former is used in the IEEE 754-standard, while the latter seems to be a bit more common in the scientific computing literature. But I have no overview how often and where mantissa is in use within the NumPy documentation. For instance, `finfo` has the attribute `nmant` which would need to be renamed, too. 

So, this topic might be a little beyond the scope of this PR.

rossbar commented on numpy/core/getlimits.py:
```diff
@@ -359,6 +360,13 @@ class finfo:
+    number in a NumPy floating point type. As in the IEEE-754 standard,
+    NumPy floating point types make use of subnormal numbers to fill the
+    gap between 0 and `tiny`. However, subnormal numbers may have a 
+    significantly reduced precision. Compare further with the Wikipedia page
+    `Denormal Numbers <https://en.wikipedia.org/wiki/Denormal_number>`_
+    
```
```suggestion
    significantly reduced precision[1]_.
    
    References
    ----------
    .. [1] Wikipedia, "Denormal Numbers",
           https://en.wikipedia.org/wiki/Denormal_number
```

rossbar replied:
Again, this is just my opinion, but I think an explicit reference reads better than a link with some connecting text (you may have to check my indentation - it's difficult to get right in the suggestion feature)

rossbar commented on numpy/core/machar.py:
```diff
@@ -41,7 +41,8 @@ class MachAr:
     xmin : float
         Floating point number ``beta**minexp`` (the smallest [in
-        magnitude] usable floating value).
+        magnitude] usable floating value with there
+        being no leading zeros in the mantissa).
```
Similar comment to my one above - rather than the "0's in the mantissa" detail, I'd prefer something like "with full precision" or "without additional loss of precision".

rossbar reviewed (commented):
Thanks for the previous changes - I have a few more suggestions, LMK what you think!

rk-mlu committed:
Update numpy/core/getlimits.py

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

rk-mlu committed:
Update getlimits: Added reference to IEEE 754 standard

rk-mlu committed:
Update of getlimits.py and machar.py

Docstrings are modified to reflect recent comments to PR.

rossbar commented on numpy/core/getlimits.py:
```diff
@@ -359,6 +358,18 @@ class finfo:
     impacts import times.  These objects are cached, so calling ``finfo()``
     repeatedly inside your functions is not a problem.
 
+    Note that `tiny` is not actually the smallest positive representable
+    value in a NumPy floating point type. As in the IEEE-754 standard[1]_,
+    NumPy floating point types make use of subnormal numbers to fill the
+    gap between 0 and `tiny`. However, subnormal numbers may have a 
+    significantly reduced precision[2]_.
```
```suggestion
    Note that ``tiny`` is not actually the smallest positive representable
    value in a NumPy floating point type. As in the IEEE-754 standard [1]_,
    NumPy floating point types make use of subnormal numbers to fill the
    gap between 0 and ``tiny``. However, subnormal numbers may have
    significantly reduced precision [2]_.
```
 * tiny should be an inline literal rather than emph
 * I think there needs to be a space before citations for numpydoc
 * grammatical nit

rk-mlu replied:
All fine with me! Many thanks!

rossbar reviewed (commented):
The changes pertaining to `Notes` and `tiny` look good to me (except for a couple minor formatting notes). However, I think something might have been lost in the changes to `minexp` - I think the original wording "power of the base" or "power of `ibeta`" was valuable. I don't mind the getting rid of the explicit use of "mantissa" in favor of the "full precision" wording, but I think the removal of the description of the exponent should be undone.

rk-mlu committed:
Update numpy/core/getlimits.py

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

rk-mlu commented:
Regarding ``minexp``: 
No problem, I will restore the original description of minexp in `finfo`and `machar`. However, in my opinion given for each attribute differs somewhat. Perhaps a further revision might be necessary to address this and the "mantissa" vs. "significand" issue.

rk-mlu committed:
restored minexp description

rk-mlu committed:
Merge branch 'finfo-doc-update' of github.com:rk-mlu/numpy into finfo-doc-update

rossbar reviewed (approved):
LGTM - Thanks @rk-mlu for the input and being so responsive!

seberg commented:
Thanks @rk-mlu for improving this and Ross for the review!
